Values & Functions :: Addressing suggestions from Issue #1762 (#1812)
ocaml/ocaml.org: data/tutorials/language/0it_00_values_functions.md

**Note**: As explained in the [Tour of OCaml](/docs/tour-of-ocaml) tutorial, the `unit` type has a single value `()`, which is pronounced "unit."

Above, the pattern does not contain any identifier, meaning no name is defined. The expression is evaluated and the side effect takes place (printing `ha ha` to standard output).

**Note**: In order for compiled files to only evaluate an expression for its side effects, you must write them after `let () =`.

<!-- user-defined single constructor variant example -->
<!-- FIXME: create an example nested pattern matching -->

### Pattern Matching on User-Defined Types

This also works with user-defined types.
```ocaml
# type live_person = int * name;;
type live_person = int * name

> # let age (years, { first, last }) = years;;

Reviewer: ```suggestion
# let age (years, { first; last }) = years;;
```

val age : live_person -> int = <fun>
```

### Discarding Values Using Pattern Matching
<!-- `_` example-->
As seen in the last example, the catch-all pattern (`_`) can be used in definitions.
```ocaml
# let (_, y) = List.split [(1, 2); (3, 4); (5, 6); (7, 8)];;
val y : int list = [2; 4; 6; 8]
```

> The `List.split` function turns a list of pairs into a pair of lists. We're only interested in the second list, we give it the name `y` and discard the first list by using `_`.

Author: ```suggestion
The `List.split` function returns a pair of lists. We're only interested in the second list, we give it the name `y` and discard the first list by using `_`.
```

**Note**: As explained in the [Tour of OCaml](/docs/tour-of-ocaml) tutorial, the `unit` type has a single value `()`, which is pronounced "unit."

Above, the pattern does not contain any identifier, meaning no name is defined. The expression is evaluated and the side effect takes place (printing `ha ha` to standard output).

> **Note**: In compiled files to evaluate an expression only for its side effects you must write them after `let () =`.

Author: ```suggestion
**Note**: In order for compiled files to only evaluate an expression for its side effects, you must write them after `let () =`.
```

<!-- user-defined single constructor variant example -->
<!-- FIXME: create an example nested pattern matching -->

> ### Pattern Matching on User-defined types

Author: ```suggestion
### Pattern Matching on User-Defined Types
```

This also works with user-defined types.
```ocaml
# type live_person = int * name;;
type live_person = int * name

> # let age (years, { first, last }) = years;;

Author: I get this error: `Error: Syntax error: '}' expected, the highlighted '{' might be unmatched`

val age : live_person -> int = <fun>
```

### Discarding Values Using Pattern Matching
<!-- `_` example-->
As seen in the last example, the catch-all pattern (`_`) can be used in definitions.
```ocaml
# let (_, y) = List.split [(1, 2); (3, 4); (5, 6); (7, 8)];;
val y : int list = [2; 4; 6; 8]
```

> The `List.split` function turns a list of pairs into a pair of lists. We're only interested in the second list, we give it the name `y` and discard the first list by using `_`.

Author: This first sentence is also on li.141

Here, the global environment is unchanged:
```ocaml
# let ten = 10 in 2 * ten;;
- : int = 20

# ten;;
Error: Unbound value ten
```

Evaluating `ten` results in an error because it hasn't been added to the global environment. However, in the expression `2 * ten`, the local environment contains the definition of `ten`.

Although OCaml is an expression-oriented language, it has a few statements. The global `let` modifies the global environment by adding a name-value *binding*.

Top-level expressions are also statements because they are equivalent to `let _ =` definitions.
```ocaml
# (1.0 +. sqrt 5.0) /. 2.0;;
- : float = 1.6180339887498949

# let _ = (1.0 +. sqrt 5.0) /. 2.0;;
- : float = 1.6180339887498949
```
<!--
With respect to the environment, there are no means to:
- List its contents
- Clear its contents
- Remove a definition
- Reset it to an earlier state
-->
### Inner Shadowing

Once you create a name, define it, and bind it to a value, it does not change. That said, a name can be defined again to create a new binding:
```ocaml
# let i = 21;;
val i : int = 21

# let i = 7 in i * 2;;
- : int = 14

> # i;;

Author: I get this error here: `Error: Unbound value i`

- : unit = ()
```

<!-- user-defined single constructor variant example -->
<!-- FIXME: create an example nested pattern matching -->
This also works with user-defined types.
```ocaml
# type live_person = int * name;;
type live_person = int * name

# let age ((years, _) : live_person) = years
val age : live_person -> int = <fun>

> ```

Author: For these I get: 

```
# type live_person = int * name;;
Error: Unbound value live_person

# let age ((years, _) : live_person) = years;;
Error: Unbound type constructor live_person
```

Also, `;;` missing at the end of `let age...`

Reviewer: Is the `name` type in scope? It's defined above.

Author: Ah! I bet that's it!

Author: That was it! Thank you!

# let h = 7;;
val h : int = 7

# e;;
- : int = 42
```

> There's now two definitions of `h` in the environment. The first `h` is unchanged. When the second `h` is defined, the first one becomes unreachable. <!--This means anything written after the second definition uses its value, but functions written *before* the second definition still use the first, even if called later.-->

Author: ```suggestion
There are now two definitions of `h` in the environment. The first `h` is unchanged. When the second `h` is defined, the first one becomes unreachable. <!--This means anything written after the second definition uses its value, but functions written *before* the second definition still use the first, even if called later.-->
```

## Defining Global Functions

> You can bind a function globally to a name using a global definition.

Author: ```suggestion
You can globally bind a function to a name using a global definition.
```

```ocaml
# let f = fun x -> x * x;;
val f : int -> int = <fun>
```

> The expression, which happens to be a function, is turned into value and bound to a name. Here is another way to do the same thing

Author: ```suggestion
The expression, which happens to be a function, is turned into value and bound to a name. Here is another way to do the same thing:
```

- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
```

This is a classic (and very inefficient) way to compute [Fibonacci](https://en.wikipedia.org/wiki/Fibonacci_sequence) numbers. The number of recursive calls created doubles at each call, which creates exponential growth.

> In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create a recursive function and recursive functions can't be anonymous.

Author: ```suggestion
In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create a recursive function, and recursive functions can't be anonymous.
```

val sweet_kitty : string -> string = <fun>
```

Since a multiple-argument function is a series of nested single-argument functions, you don't have to pass all arguments at once.

Passing a single parameter to `sour_kitty` or `sweet_kitty` returns a function of type `string -> string`. 
 The first argument, here `"kitty"`, is captured and the result is a [closure](#closures).

These expressions have the same value:
- `fun x -> sweet_cat "kitty" x`
- `sweet_cat "kitty"`

### Types of Functions of Multiple Argument

> Let's look at the types in this definition:

Author: ```suggestion
Let's look at the types here:
```
